### PR TITLE
Rework Bubble Sort C++

### PR DESCRIPTION
**Pull request by benchislett**
Bring on the criticism!

### CODE REVIEW

File: chapters/algorithms/bubble_sort/code/c++/bubblesort.cpp [hunk 1]
```diff
@@ -1,71 +1,39 @@
-#include <algorithm>
-#include <cstddef>
-#include <iostream>
-#include <iterator>
-#include <random>
+#include <stdlib.h> // rand
```
Comment (antonte) [inline chapters/algorithms/bubble_sort/code/c++/bubblesort.cpp]: do
```C++
#include <cstdlib>
```

File: chapters/algorithms/bubble_sort/code/c++/bubblesort.cpp [hunk 2]
```diff
@@ -1,71 +1,39 @@
...
-
-template <class Iter>
-void bubble_sort(Iter const first, Iter const last) {
-  if (first == last) {
-    // the empty range is vacuously sorted
-    return;
-  }
-
-  for (;;) {
+std::vector<int> bubble_sort(std::vector<int> unsorted_list) {
```
Comment (antonte) [inline chapters/algorithms/bubble_sort/code/c++/bubblesort.cpp]: I would do sort in place
```C++
void bubble_sort(std::vector<int>& array)
```

File: chapters/algorithms/bubble_sort/code/c++/bubblesort.cpp [hunk 3]
```diff
@@ -1,71 +1,39 @@
...
 }
 
-int main() {
-  std::random_device random_device;
-  auto rng = std::mt19937(random_device());
-
-  auto input = generate_input(10, rng);
+int main()
+{
+  std::vector<int> to_be_sorted = {};
```
Comment (antonte) [inline chapters/algorithms/bubble_sort/code/c++/bubblesort.cpp]: style minor: `= {}` just a noise, do:
```c++
std::vector<int> to_be_sorted;
```

File: chapters/algorithms/bubble_sort/code/c++/bubblesort.cpp [hunk 4]
```diff
@@ -1,71 +1,39 @@
...
+  // Initialize and print a vector with 50 random integers of domain [0,1000]
+  for (int i = 0; i < 50; i++) {
```
Comment (antonte) [inline chapters/algorithms/bubble_sort/code/c++/bubblesort.cpp]: style minor: ++i is preferable style. 

https://stackoverflow.com/questions/24853/what-is-the-difference-between-i-and-i

File: chapters/algorithms/bubble_sort/code/c++/bubblesort.cpp [hunk 5]
```diff
@@ -1,71 +1,39 @@
...
+    int value = rand() % 1000;
```
Comment (antonte) [inline chapters/algorithms/bubble_sort/code/c++/bubblesort.cpp]: You may want to initialize random generator with 
```c++
srand(time(nullptr));
```

File: chapters/algorithms/bubble_sort/code/c++/bubblesort.cpp [hunk 6]
```diff
@@ -1,71 +1,39 @@
...
+    to_be_sorted.push_back(value);
+  }
+  std::cout << std::endl;
 
-  std::cout << "before sorting:\n";
-  print_range(std::cout, begin(input), end(input));
+  std::vector<int> sorted = bubble_sort(to_be_sorted);
 
-  bubble_sort(begin(input), end(input));
+  for (int i : sorted) { std::cout << i << " "; }
```
Comment (antonte) [inline chapters/algorithms/bubble_sort/code/c++/bubblesort.cpp]: add 
```c++
std::cout << std::endl; 
```
It is kind of annoying if terminal program does not do it.

File: chapters/algorithms/bubble_sort/code/c++/bubblesort.cpp [hunk 7]
```diff
@@ -1,71 +1,39 @@
...
 
-  std::cout << "\nafter sorting:\n";
-  print_range(std::cout, begin(input), end(input));
+  return 0;
```
Comment (antonte) [inline chapters/algorithms/bubble_sort/code/c++/bubblesort.cpp]: style minor: C++ does not require return 0

File: chapters/algorithms/bubble_sort/code/c++/bubblesort.cpp [hunk 8]
```diff
@@ -1,71 +1,39 @@
...
+  while (true) {
     bool is_sorted = true;
-
-    for (auto it = first; it + 1 != last; ++it) {
-      // these are unsorted! gotta swap 'em
-      if (*(it + 1) < *it) {
-        using std::swap;
-        swap(*it, *(it + 1));
+    // Sweep through the array
+    for (unsigned int i = 0; i < unsorted_list.size() - 1; i++) {
```
Comment (antonte) [inline chapters/algorithms/bubble_sort/code/c++/bubblesort.cpp]: style minor: ++i is preferable style.

https://stackoverflow.com/questions/24853/what-is-the-difference-between-i-and-i

File: chapters/algorithms/bubble_sort/code/c++/bubblesort.cpp [hunk 9]
```diff
@@ -1,71 +1,39 @@
...
 
-int main() {
-  std::random_device random_device;
-  auto rng = std::mt19937(random_device());
-
-  auto input = generate_input(10, rng);
+int main()
+{
+  std::vector<int> to_be_sorted = {};
+  // Initialize and print a vector with 50 random integers of domain [0,1000]
```
Comment (antonte) [inline chapters/algorithms/bubble_sort/code/c++/bubblesort.cpp]: Comment is a lie, should be:
```
// Initialize and print a vector with 50 random integers of domain [0,999]
```
or
```
// Initialize and print a vector with 50 random integers of domain [0,1000)
```
First is preferable.

File: chapters/algorithms/bubble_sort/code/c++/bubblesort.cpp (hunk 4 above)
Comment (antonte) [inline chapters/algorithms/bubble_sort/code/c++/bubblesort.cpp]: I would provide a hand written test sample. Random test set usually does not cover edge cases and gives false confidence about the code.

In this case it does not cover cases with empty array, already sorted array and array with reverse order for instance.

**Comment by antonte** [comment]
Do you need to update https://github.com/algorithm-archivists/algorithm-archive/blob/master/chapters/algorithms/bubble_sort/bubble_sort.md ?

**Comment by Butt4cak3** [comment]
I don't really have any criticism for your code, but I can tell you that we already have an open pull request about Bubble Sort in C++, which is #244. We can't possible merge both of them because they will certainly cause merge conflicts.

Because we use a "first come - first serve" approach, I'm going to have to close this issue. I'm sorry for that, @benchislett.

We still appreciate your contribution, though! I hope that you will stay around and find something else you can do for the Algorithm Archive. Just next time, make sure that there's not already a PR that does the same thing. ;)
